endpoints/service.go

```diff
@@ -175,38 +181,50 @@ func (m *serviceMap) register(srv interface{}, name, ver, desc string, isDefault
 // It doesn't create ServiceMethod.info if internal == true
 func newServiceMethod(m *reflect.Method, internal bool) *ServiceMethod {
 	mtype := m.Type
+	in, out := mtype.NumIn(), mtype.NumOut()
 	// Method must be exported.
-	// Method needs four ins: receiver, *http.Request, *args, *reply.
-	if m.PkgPath != "" || mtype.NumIn() != 4 {
+	if m.PkgPath != "" {
 		return nil
 	}
-	// First argument must be a pointer and must be http.Request.
-	reqType := mtype.In(1)
-	if reqType.Kind() != reflect.Ptr || reqType.Elem() != typeOfRequest {
+	// Method needs 3 ints and 2 outs, or 4 ins and 1 out:
+	// receiver, *http.Request|Context, *first, *res1 => error
+	// receiver, *http.Request|Context, *first        => *res1, error
+	if !(in == 3 && out == 2 || in == 4 && out == 1) {
```

Comment:
I think it's better to use a switch statement here for clarity. I would write:

	var httpReqType, reqType, respType, errType reflect.Type
	mtype := m.Type
	switch {
	// method(receiver, *http.Request, *reqType, *resType) error
	case mtype.NumIn() == 4 && mtype.NumOut() == 1:
		httpReqType = mtype.In(1)
		reqType = mtype.In(2)
		respType = mtype.In(3)
		errType = mtype.Out(0)
	// method(receiver, *http.Request, *reqType) (*resType, error)
	case mtype.NumIn() == 3 && mtype.NumOut() == 2:
		httpReqType = mtype.In(1)
		reqType = mtype.In(2)
		respType = mtype.Out(0)
		errType = mtype.Out(1)
	default:
		return nil
	}